File: cores/java/src/main/java/nes/core/Joypad.java
package nes.core;

// NES joypad state for two standard controllers (port of cores/csharp/src/Joypad.cs).
public final class Joypad {
    public static final int CONTROLLER_COUNT = 2;
    public static final int BUTTON_COUNT = 8;

    public static final int BUTTON_A = 0;
    public static final int BUTTON_B = 1;
    public static final int BUTTON_SELECT = 2;
    public static final int BUTTON_START = 3;
    public static final int BUTTON_UP = 4;
    public static final int BUTTON_DOWN = 5;
    public static final int BUTTON_LEFT = 6;
    public static final int BUTTON_RIGHT = 7;

    public final byte[] current = new byte[CONTROLLER_COUNT];
    public boolean strobe;
    public final byte[] shift = new byte[CONTROLLER_COUNT];
    public final byte[] counter = new byte[CONTROLLER_COUNT];

    public Joypad() {}

    public void setButton(int controller, int button, boolean pressed) {
        if (controller < 0 || controller >= CONTROLLER_COUNT || button < 0 || button >= BUTTON_COUNT)
            return;
        int mask = 1 << button;
        if (pressed) current[controller] |= mask;
        else current[controller] &= ~mask;
        if (strobe) shift[controller] = current[controller];
    }

    public void writeStrobe(int value) {
        boolean newStrobe = (value & 0x01) != 0;
        if (strobe && !newStrobe) {
            for (int c = 0; c < CONTROLLER_COUNT; ++c) {
                shift[c] = current[c];
                counter[c] = 0;
            }
        }
        strobe = newStrobe;
        if (strobe) {
            for (int c = 0; c < CONTROLLER_COUNT; ++c)
                shift[c] = current[c];
        }
    }

    public int read(int controller) {
        if (controller < 0 || controller >= CONTROLLER_COUNT)
            return 1;
        if (strobe)
            return current[controller] & 0x01;
        int bit;
        if ((counter[controller] & 0xFF) < BUTTON_COUNT)
            bit = (shift[controller] >> (counter[controller] & 0xFF)) & 0x01;
        else
            bit = 1;
        if ((counter[controller] & 0xFF) < 0xFF)
            counter[controller] = (byte)((counter[controller] & 0xFF) + 1);
        return bit;
    }

    public void clear() {
        for (int c = 0; c < CONTROLLER_COUNT; ++c) {
            current[c] = 0;
            if (strobe) shift[c] = 0;
        }
    }
}

File: cores/go/internal/joypad/joypad.go
// Package joypad implements the NES standard controller input
// (port of cores/c/src/joypad.c).
package joypad

import "unsafe"

const (
	ControllerCount = 2
	ButtonCount     = 8
)

// Button bit indices.
const (
	ButtonA      = 0
	ButtonB      = 1
	ButtonSelect = 2
	ButtonStart  = 3
	ButtonUp     = 4
	ButtonDown   = 5
	ButtonLeft   = 6
	ButtonRight  = 7
)

// Joypad is the NES joypad state for two standard controllers.
type Joypad struct {
	Current [ControllerCount]uint8
	Strobe  bool
	Shift   [ControllerCount]uint8
	Counter [ControllerCount]uint8
}

// Init constructs a joypad with no buttons pressed and the strobe low.
func (j *Joypad) Init() {
	*j = Joypad{}
}

// SetButton sets or clears a button on the given controller.
func (j *Joypad) SetButton(controller, button uint8, pressed bool) {
	if controller >= ControllerCount || button >= ButtonCount {
		return
	}
	mask := uint8(1 << button)
	if pressed {
		j.Current[controller] |= mask
	} else {
		j.Current[controller] &= ^mask
	}
	if j.Strobe {
		j.Shift[controller] = j.Current[controller]
	}
}

// WriteStrobe writes to $4016 bit 0 - the controller strobe line.
func (j *Joypad) WriteStrobe(value uint8) {
	newStrobe := (value & 0x01) != 0
	if j.Strobe && !newStrobe {
		for c := uint8(0); c < ControllerCount; c++ {
			j.Shift[c] = j.Current[c]
			j.Counter[c] = 0
		}
	}
	j.Strobe = newStrobe
	if j.Strobe {
		for c := uint8(0); c < ControllerCount; c++ {
			j.Shift[c] = j.Current[c]
		}
	}
}

// Read reads one bit from the given controller.
func (j *Joypad) Read(controller uint8) uint8 {
	if controller >= ControllerCount {
		return 1
	}
	if j.Strobe {
		return j.Current[controller] & 0x01
	}
	var bit uint8
	if j.Counter[controller] < ButtonCount {
		bit = (j.Shift[controller] >> j.Counter[controller]) & 0x01
	} else {
		bit = 1
	}
	if j.Counter[controller] < 0xFF {
		j.Counter[controller]++
	}
	return bit
}

// StrobeState returns the current strobe line state.
func (j *Joypad) StrobeState() bool { return j.Strobe }

// Clear clears all buttons on both controllers.
func (j *Joypad) Clear() {
	for c := uint8(0); c < ControllerCount; c++ {
		j.Current[c] = 0
		if j.Strobe {
			j.Shift[c] = 0
		}
	}
}

// CurrentState returns live button state for a controller.
func (j *Joypad) CurrentState(controller uint8) uint8 {
	if controller >= ControllerCount {
		return 0
	}
	return j.Current[controller]
}

// ShiftState returns the frozen snapshot for a controller.
func (j *Joypad) ShiftState(controller uint8) uint8 {
	if controller >= ControllerCount {
		return 0
	}
	return j.Shift[controller]
}

// SerializedSize returns the byte length of the joypad state serialization.
func SerializedSize() int { return int(unsafe.Sizeof(Joypad{})) }

// Serialize writes the joypad state into buf (must be >= SerializedSize).
// Returns bytes written.
func Serialize(j *Joypad, buf []byte) int {
	n := SerializedSize()
	if len(buf) < n {
		return 0
	}
	src := (*[1 << 30]byte)(unsafe.Pointer(j))[:n:n]
	copy(buf, src)
	return n
}

// Deserialize restores the joypad state from buf. Returns bytes read.
func Deserialize(j *Joypad, buf []byte) int {
	n := SerializedSize()
	if len(buf) < n {
		return 0
	}
	dst := (*[1 << 30]byte)(unsafe.Pointer(j))[:n:n]
	copy(dst, buf)
	return n
}
